### ISSUE TITLE
Packer cannot copy scripts to VMs `/tmp` folder

### ISSUE THREAD

**Issue by Wasapl** [hashicorp/packer#6151]
Packer cannot scp script to /tmp folder.
Next line shows up in log: "scp output: NUL scp: Broken pipe." Then packer tries to scp script again and again until timeout. 
/tmp is writtable, with sufficient amount of free memory. 
If I run command `scp basicconfig.sh vagrant@<IP>:/tmp` it works like a charm.

- Packer version 1.2.2
- Host platform Windows Server 2016
- Debug log https://gist.github.com/Wasapl/9e871d22f88b9bd5380ab65daa6e995d
- The _simplest example template and scripts_ needed to reproduce the bug: https://gist.github.com/Wasapl/720ab8a16c4d1b4a45bc8b92639f298f contents of provisioned scripts are irrelevant.

**Comment by krisztianloki** [comment]
This is a regression from 1.2.1
It seems to have something to do with windows-unix line ending conversion; setting the *binary* option to True fixes the upload problem in 1.2.2.

**Comment by Wasapl** [comment]
Thanks, @krisztianloki ! Downgraded to 1.2.1 and now it works.

**Comment by SwampDragons** [comment]
@krisztianloki Thanks for the workaround -- do you have any more details on the specifics of the issue or ideas on how we can fix it for 1.2.3?

**Comment by krisztianloki** [comment]
I think that it has nothing to do with any specific builder; it happened to me with the virtualbox-ovf builder.
Here is the relevant part from the debug log:
>
2018/04/23 10:55:45 [INFO] (telemetry) Starting provisioner shell
2018/04/23 10:55:45 ui: ==> virtualbox-ovf: Provisioning with shell script: scripts/create_rootdisk
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 Opening scripts/create_rootdisk for reading
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [DEBUG] Opening new ssh session
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [DEBUG] starting remote command: test -d "/tmp/script_9991.sh"
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [INFO] 5903 bytes written for 'uploadData'
2018/04/23 10:55:45 [INFO] 5903 bytes written for 'uploadData'
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [ERROR] Remote command exited with '1': test -d "/tmp/script_9991.sh"
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [DEBUG] Opening new ssh session
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [DEBUG] Starting remote scp process:  scp -vt /tmp
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [DEBUG] Started SCP session, beginning transfers...
2018/04/23 10:55:45 packer.exe: 2018/04/23 10:55:45 [DEBUG] scp: Uploading script_9991.sh: perms=C0666 size=6112
2018/04/23 10:55:46 packer.exe: 2018/04/23 10:55:46 [DEBUG] SCP session complete, closing stdin pipe.
2018/04/23 10:55:46 packer.exe: 2018/04/23 10:55:46 [DEBUG] Waiting for SSH session to complete.
2018/04/23 10:55:46 packer.exe: 2018/04/23 10:55:46 [DEBUG] non-zero exit status: 1
2018/04/23 10:55:46 packer.exe: 2018/04/23 10:55:46 [DEBUG] scp output:  ☺scp: Broken pipe
2018/04/23 10:55:46 packer.exe: 2018/04/23 10:55:46 Retryable error: Error uploading script: Process exited with status 1
>
The real size of the file is 6112 bytes with windows line-ending, packer converts it to linux format and the result is a 5903 byte long file. But this information gets lost somewhere and scp thinks it has to upload the original size. The difference between 6112 and 5903 is exactly the number of lines in the file.

I have no idea how to fix it, I am not familiar with Go... I was looking at the code trying to figure out something but got lost...

**Comment by breisig** [comment]
I just tried with packer 1.2.3 and I am still getting the same issue with virtualbox-ovf builder

**Comment by SwampDragons** [comment]
@breisig can I have debug logs from your failed build? And a repro case, if you've got a simple one?

**Comment by andrewsav-datacom** [comment]
@SwampDragons where you able to reproduce this one? I'm also affected, and I can try and put a repro together if you still have not got one.

**Comment by breisig** [comment]
@SwampDragons Here's the log output using virtualbox-ovf on packer 1.2.3 : https://pastebin.com/siFdFTwA

**Comment by SwampDragons** [comment]
@andrepearce I'd love a repro, thanks.

**Comment by SwampDragons** [comment]
@andrewsav-datacom I'd love a repro, thanks

**Comment by andrewsav-datacom** [comment]
[ubuntu-1804.zip](https://github.com/hashicorp/packer/files/1962335/ubuntu-1804.zip)
@SwampDragons this is not quite minimal but it reproduces it for me. It works on 1.2.1 but fails on 1.2.3 with the error in the OP. Note that in my case I'm not using Hyper-V.

**Comment by andrewsav-datacom** [comment]
[ubuntu-1604.zip](https://github.com/hashicorp/packer/files/1962500/ubuntu-1604.zip) Here us repo for VirtualBox/VmWare , the logs are inside.

**Comment by AndrewSav** [comment]
@SwampDragons yes!

**Comment by SwampDragons** [comment]
okay; I am able to repro with @andrewsav-datacom 's case, but I'm not able to fix the issue by adding the "binary" flag as @krisztianloki suggested. Downgrading to Packer 1.2.1 did solve it though, so I am confident this is the same bug. I'm going to start combing through the commit history and building/testing on commits I think could be related.

**Comment by SwampDragons** [comment]
andddd the culprit is: c5e0710a9a1c9d99a8241bc10ec21f3766e3719c

I'll start looking at fixes tomorrow.

**Comment by SwampDragons** [comment]
@andrewsav-datacom I have a patch that seems to resolve the issue with your use case:  https://github.com/hashicorp/packer/pull/6239. A windows-build of said patch is attached.
[packer.zip](https://github.com/hashicorp/packer/files/1972104/packer.zip) . Can you and @breisig test with this patch and let me know if it resolves your issues?  I wonder if something in the act of stating the file is causing weirdness with the endings. I'm still going to dig in a bit further but would like confirmation that I'm on the right track.

**Comment by breisig** [comment]
@SwampDragons Thanks, I'll test it out right away.

**Comment by SwampDragons** [comment]
Glad to hear it!  I'll make sure this is fixed in one capacity or another in the next release.

**Comment by hegyre** [comment]
Hello @SwampDragons ,

Had the exact same issue with Packer 1.2.2 and 1.2.3, and your dev build fixed the issue for me as well ;)

**Comment by SwampDragons** [comment]
yeah, I saw that... I'm just really confused about how stating the file could have messed with the EOL stuff. I'll dig into it later. For now I'm glad you have a workaround.

**Comment by krisztianloki** [comment]
I think the EOL conversion happens as part of the uploading. And if the conversion is not an identity conversion, then the size of the file will change. I assume that stat() happens before the conversion; that is why the upload fails with size mismatch.

**Comment by lmayorga1980** [comment]
Same issue here using `git bash` on Windows.

**Comment by AndrewSav** [comment]
@staticfrost, most likely it is _already_ in the next release. All we need to do is to wait for that release to come out.

**Comment by SwampDragons** [comment]
That's correct. We'll be releasing 1.3.0 either at the end of this week or the beginning of next.
